File: go/sigil/client_tags_test.go
package sigil

import (
	"context"
	"testing"
	"time"

	"go.opentelemetry.io/otel/attribute"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/metric/metricdata"
)

const clientTagProjectKey = spanAttrTagPrefix + "project"

func TestClientTagsOnGenerationSpanAndMetrics(t *testing.T) {
	metricReader := sdkmetric.NewManualReader()
	meterProvider := sdkmetric.NewMeterProvider(sdkmetric.WithReader(metricReader))
	t.Cleanup(func() {
		_ = meterProvider.Shutdown(context.Background())
	})

	client, recorder, _ := newTestClient(t, Config{
		Tags:  map[string]string{"project": "checkout-svc"},
		Meter: meterProvider.Meter("sigil-test"),
		Now: func() time.Time {
			return time.Date(2026, 6, 4, 12, 0, 0, 0, time.UTC)
		},
	})

	_, genRec := client.StartGeneration(context.Background(), GenerationStart{
		ConversationID: "conv-tags",
		Model:          ModelRef{Provider: "openai", Name: "gpt-5"},
	})
	genRec.SetResult(Generation{
		Usage: TokenUsage{InputTokens: 10, OutputTokens: 5},
	}, nil)
	genRec.End()

	span := onlyGenerationSpan(t, recorder.Ended())
	attrs := spanAttributeMap(span)
	if got := attrs[clientTagProjectKey].AsString(); got != "checkout-svc" {
		t.Fatalf("expected %s=checkout-svc on generation span, got %q", clientTagProjectKey, got)
	}

	collected := collectMetrics(t, metricReader)
	_ = findHistogramPointForTags(t, findHistogram(t, collected, metricOperationDuration), map[string]string{
		clientTagProjectKey: "checkout-svc",
	})
}

func TestClientTagsOnEmbeddingAndToolSpans(t *testing.T) {
	client, recorder, _ := newTestClient(t, Config{
		Tags: map[string]string{"project": "embed-tools"},
	})

	_, embedRec := client.StartEmbedding(context.Background(), EmbeddingStart{
		Model: ModelRef{Provider: "openai", Name: "text-embedding-3-small"},
	})
	embedRec.SetResult(EmbeddingResult{InputTokens: 1})
	embedRec.End()

	embedSpan := onlyEmbeddingSpan(t, recorder.Ended())
	if got := spanAttributeMap(embedSpan)[clientTagProjectKey].AsString(); got != "embed-tools" {
		t.Fatalf("expected %s on embedding span, got %q", clientTagProjectKey, got)
	}

	_, toolRec := client.StartToolExecution(context.Background(), ToolExecutionStart{
		ToolName: "weather",
	})
	toolRec.SetResult(ToolExecutionEnd{Result: map[string]any{"temp": 72}})
	toolRec.End()

	toolSpan := onlyToolSpan(t, recorder.Ended())
	if got := spanAttributeMap(toolSpan)[clientTagProjectKey].AsString(); got != "embed-tools" {
		t.Fatalf("expected %s on tool span, got %q", clientTagProjectKey, got)
	}
}

func TestClientTagsEmptyIsNoOpOnSpan(t *testing.T) {
	client, recorder, _ := newTestClient(t, Config{})

	_, genRec := client.StartGeneration(context.Background(), GenerationStart{
		Model: ModelRef{Provider: "openai", Name: "gpt-5"},
	})
	genRec.SetResult(Generation{}, nil)
	genRec.End()

	span := onlyGenerationSpan(t, recorder.Ended())
	attrs := spanAttributeMap(span)
	if _, ok := attrs[clientTagProjectKey]; ok {
		t.Fatalf("did not expect %s when client tags are unset", clientTagProjectKey)
	}
}

func TestPerCallGenerationTagsStayExportOnly(t *testing.T) {
	exporter := &capturingGenerationExporter{}
	client, recorder, _ := newTestClient(t, Config{
		testGenerationExporter: exporter,
	})

	_, genRec := client.StartGeneration(context.Background(), GenerationStart{
		Model: ModelRef{Provider: "openai", Name: "gpt-5"},
		Tags:  map[string]string{"call_only": "yes"},
	})
	genRec.SetResult(Generation{}, nil)
	genRec.End()

	span := onlyGenerationSpan(t, recorder.Ended())
	if _, ok := spanAttributeMap(span)[spanAttrTagPrefix+"call_only"]; ok {
		t.Fatalf("per-call tag must not appear on span without client-level SIGIL_TAGS")
	}

	if err := client.Flush(context.Background()); err != nil {
		t.Fatalf("flush: %v", err)
	}

	if len(exporter.requests) != 1 || len(exporter.requests[0].Generations) != 1 {
		t.Fatalf("expected one exported generation, got %#v", exporter.requests)
	}
	if got := exporter.requests[0].Generations[0].GetTags()["call_only"]; got != "yes" {
		t.Fatalf("expected call_only=yes on export tags, got %#v", exporter.requests[0].Generations[0].GetTags())
	}
}

func collectMetrics(t *testing.T, reader *sdkmetric.ManualReader) metricdata.ResourceMetrics {
	t.Helper()
	var collected metricdata.ResourceMetrics
	if err := reader.Collect(context.Background(), &collected); err != nil {
		t.Fatalf("collect metrics: %v", err)
	}
	return collected
}

func findHistogramPointForTags(t *testing.T, histogram metricdata.Histogram[float64], want map[string]string) metricdata.HistogramDataPoint[float64] {
	t.Helper()
	for _, point := range histogram.DataPoints {
		if histogramPointMatchesForTags(point.Attributes, want) {
			return point
		}
	}
	t.Fatalf("expected histogram point with attrs %v", want)
	return metricdata.HistogramDataPoint[float64]{}
}

func histogramPointMatchesForTags(attrs attribute.Set, want map[string]string) bool {
	for key, expected := range want {
		value, ok := (&attrs).Value(attribute.Key(key))
		if !ok || value.AsString() != expected {
			return false
		}
	}
	return true
}

func findHistogram(t *testing.T, collected metricdata.ResourceMetrics, metricName string) metricdata.Histogram[float64] {
	t.Helper()
	for _, scopeMetrics := range collected.ScopeMetrics {
		for _, metric := range scopeMetrics.Metrics {
			if metric.Name != metricName {
				continue
			}
			histogram, ok := metric.Data.(metricdata.Histogram[float64])
			if !ok {
				t.Fatalf("metric %q is not a float64 histogram", metricName)
			}
			return histogram
		}
	}
	t.Fatalf("metric %q not found", metricName)
	return metricdata.Histogram[float64]{}
}

File: js/test/client.tags.test.mjs
import assert from 'node:assert/strict';
import test from 'node:test';
import {
  AggregationTemporality,
  DataPointType,
  InMemoryMetricExporter,
  MeterProvider,
  PeriodicExportingMetricReader,
} from '@opentelemetry/sdk-metrics';
import { BasicTracerProvider, InMemorySpanExporter, SimpleSpanProcessor } from '@opentelemetry/sdk-trace-base';
import { Agento11yClient, defaultConfig } from '../.test-dist/index.js';

const clientTagProjectKey = 'agento11y.tag.project';

class CapturingExporter {
  requests = [];

  async exportGenerations(request) {
    this.requests.push(structuredClone(request));
    return {
      results: request.generations.map((generation) => ({
        generationId: generation.id,
        accepted: true,
      })),
    };
  }
}

test('client SIGIL_TAGS appear on generation span and metrics', async () => {
  const harness = newHarness({ tags: { project: 'checkout-svc' } });

  try {
    const recorder = harness.client.startGeneration({
      model: { provider: 'openai', name: 'gpt-5' },
    });
    recorder.setResult({
      usage: { inputTokens: 10, outputTokens: 5 },
    });
    recorder.end();

    const span = singleGenerationSpan(harness.spanExporter);
    assert.equal(span.attributes[clientTagProjectKey], 'checkout-svc');

    const metricAttrs = await harness.metricDataPointAttributes('gen_ai.client.operation.duration');
    assert.ok(
      metricAttrs.some((attrs) => attrs[clientTagProjectKey] === 'checkout-svc'),
      'expected agento11y.tag.project on operation.duration metric',
    );
  } finally {
    await shutdownHarness(harness);
  }
});

test('client tags on embedding and tool spans', async () => {
  const harness = newHarness({ tags: { project: 'embed-tools' } });

  try {
    const embed = harness.client.startEmbedding({
      model: { provider: 'openai', name: 'text-embedding-3-small' },
    });
    embed.setResult({ inputTokens: 1 });
    embed.end();

    const embedSpan = singleEmbeddingSpan(harness.spanExporter);
    assert.equal(embedSpan.attributes[clientTagProjectKey], 'embed-tools');

    const tool = harness.client.startToolExecution({ toolName: 'weather' });
    tool.setResult({ result: 'sunny' });
    tool.end();

    const toolSpan = singleToolSpan(harness.spanExporter);
    assert.equal(toolSpan.attributes[clientTagProjectKey], 'embed-tools');
  } finally {
    await shutdownHarness(harness);
  }
});

test('empty client tags are omitted from spans', async () => {
  const harness = newHarness();

  try {
    const recorder = harness.client.startGeneration({
      model: { provider: 'openai', name: 'gpt-5' },
    });
    recorder.end();

    const span = singleGenerationSpan(harness.spanExporter);
    assert.equal(span.attributes[clientTagProjectKey], undefined);
  } finally {
    await shutdownHarness(harness);
  }
});

test('per-call generation tags stay export-only', async () => {
  const harness = newHarness();

  try {
    const recorder = harness.client.startGeneration({
      model: { provider: 'openai', name: 'gpt-5' },
      tags: { call_only: 'yes' },
    });
    recorder.end();

    const span = singleGenerationSpan(harness.spanExporter);
    assert.equal(span.attributes['agento11y.tag.call_only'], undefined);

    await harness.client.shutdown();
    const generation = harness.generationExporter.requests[0]?.generations?.[0];
    assert.equal(generation?.tags?.call_only, 'yes');
  } finally {
    await harness.traceProvider.shutdown();
  }
});

function newHarness(overrides = {}) {
  const spanExporter = new InMemorySpanExporter();
  const traceProvider = new BasicTracerProvider({
    spanProcessors: [new SimpleSpanProcessor(spanExporter)],
  });
  const tracer = traceProvider.getTracer('agento11y-sdk-js-test');
  const generationExporter = new CapturingExporter();
  const defaults = defaultConfig();

  const metricExporter = new InMemoryMetricExporter(AggregationTemporality.CUMULATIVE);
  const metricReader = new PeriodicExportingMetricReader({
    exporter: metricExporter,
    exportIntervalMillis: 60_000,
  });
  const meterProvider = new MeterProvider({ readers: [metricReader] });

  const client = new Agento11yClient({
    tracer,
    meter: meterProvider.getMeter('sigil-test'),
    generationExport: {
      ...defaults.generationExport,
      batchSize: 100,
      flushIntervalMs: 60_000,
      maxRetries: 1,
      initialBackoffMs: 1,
      maxBackoffMs: 1,
    },
    generationExporter,
    ...overrides,
  });

  return {
    client,
    spanExporter,
    traceProvider,
    generationExporter,
    metricExporter,
    meterProvider,
    metricReader,
    async metricDataPointAttributes(metricName) {
      await metricReader.forceFlush();
      const metric = metricExporter
        .getMetrics()
        .flatMap((resourceMetrics) => resourceMetrics.scopeMetrics)
        .flatMap((scopeMetrics) => scopeMetrics.metrics)
        .find((m) => m.descriptor.name === metricName && m.dataPointType === DataPointType.HISTOGRAM);
      assert.ok(metric, `expected histogram metric ${metricName}`);
      return metric.dataPoints.map((point) => point.attributes);
    },
  };
}

async function shutdownHarness(harness) {
  await harness.client.shutdown();
  await harness.meterProvider.shutdown();
  await harness.traceProvider.shutdown();
}

function generationSpans(spanExporter) {
  return spanExporter.getFinishedSpans().filter((span) => {
    const operation = span.attributes['gen_ai.operation.name'];
    return operation !== 'execute_tool' && operation !== 'embeddings';
  });
}

function embeddingSpans(spanExporter) {
  return spanExporter.getFinishedSpans().filter((span) => span.attributes['gen_ai.operation.name'] === 'embeddings');
}

function toolSpans(spanExporter) {
  return spanExporter.getFinishedSpans().filter((span) => span.attributes['gen_ai.operation.name'] === 'execute_tool');
}

function singleGenerationSpan(spanExporter) {
  const spans = generationSpans(spanExporter);
  assert.equal(spans.length, 1);
  return spans[0];
}

function singleEmbeddingSpan(spanExporter) {
  const spans = embeddingSpans(spanExporter);
  assert.equal(spans.length, 1);
  return spans[0];
}

function singleToolSpan(spanExporter) {
  const spans = toolSpans(spanExporter);
  assert.equal(spans.length, 1);
  return spans[0];
}
